balancer.go

```diff
@@ -104,3 +109,44 @@ func (lb *LeastBytes) makeCounters(partitions ...int) (counters []leastBytesCoun
 	})
 	return
 }
+
+var (
+	fnv1aPool = &sync.Pool{
+		New: func() interface{} {
+			return fnv.New32a()
+		},
+	}
+)
+
+// Hash is a Balancer implementation that routes messages to the partition
+// based on the FNV-1a hash of the key.  This ensures that messages with the
+// same key end up on the same partition.
+//
+// Note that this is the same algorithm used by the Sarama Producer and ensures
+// that messages produced by kafka-go will be delivered to the same topics that
+// the Sarama producer would be delivered to
+type Hash struct {
+	rr RoundRobin
+}
+
+func (h *Hash) Balance(msg Message, partitions ...int) (partition int) {
+	if msg.Key == nil {
+		return (&h.rr).Balance(msg, partitions...)
```

Comment:
Sarama's implementation uses a random distribution if the key is `nil` (see https://github.com/Shopify/sarama/blob/8093021a9bebdd8046c88bbc2c6f747e5a03cf66/partitioner.go#L114-L116).

I don't know what's best here, round robin likely would give a better balancing but would also not be exactly the same behavior as Sarama's so switching to kafka-go would result in a behavior change.

We should document the behavior when the key is `nil` and if it's different from the one implemented in Sarama we should also make it clear (since the documentation currently claims that it's the same algorithm).

Another alternative is to allow the program to configure the fallback balancer in the `Hash` structure by exporting the field.

Reply:
True, but since the random distribution isn't something that can be counted on by consumer, I felt that using one of the Balancers that kafka-go already provided was acceptable.  The driving issue for me was having interoperability between sarama and kafka-go.

Happy to create a random balancer if you feel this is important.

Reply:
The use case I was using as a driver is, if I had a heterogeneous environment with both kafka-go and sarama would I be surprised.  If I set the message key, then I expect messages with the same key to arrive on the same partition.  If I don't set the key, then I have no expectation as to which partition the message will arrive.

Reply:
You're right, random cannot be relied on, and round robin will give a more balanced distribution. Let's keep it this way 👍